### ISSUE TITLE
Should alignments methods be abstract?

### ISSUE THREAD

**Issue by h-mayorquin** [catalystneuro/neuroconv#439]
Currently, the `BaseDataInterface` requires child classes to implement methods related to alignment, as outlined in the following link:

https://github.com/catalystneuro/neuroconv/blob/5c6b5d982bc896c5ec05b9cc417f5040e4d967ef/src/neuroconv/basedatainterface.py#L44-L90

This however,  breaks compatibility with previous conversion code, as behavior interfaces often did not implement these methods, causing them to fail under the current contract.

However, the second point could be seen as a "feature, not a bug" situation. As @CodyCBakerPhD suggests, the intention may be to encourage developers to think carefully about these issues and implement empty methods with a pass statement if no further action is needed. Both points have their merits.

To summarize, the current approach has its pros and cons:

Pros: 
* Encourages developers to think about alignmen from the start.
* Can be bypassed easily by defining an empty method.
Cons:
* Breaks compatibility with previous code.
* Introduces more frictions in development.  It increases development time by imposing additional requirements on developers. In other words, anyone creating a new interface now needs to implement these methods and consider their implications, which incurs an upfront cost.

@CodyCBakerPhD points out that past conversions should have requirements that allow them to run with those specific requirements. However, it is uncertain how thorough the project has been in implementing these requirements, leaving some doubt as to their effectiveness.

**Comment by bendichter** [comment]
Thanks @h-mayorquin , this is a great summary.

For me, I would prefer a solution that does not break backwards compatibility. For example, if a lab wanted to update their conversion to use the new Bruker interface, or any interface we develop in the future, this would require them to upgrade their neuroconv version, which would break all of their existing custom interfaces. While we don't necessarily guarantee that labs will be able to upgrade, I think it makes sense to try to adopt a development pattern that allows them to do so.

For me, a `NotImplementedError` when a user tried to time-align is a pretty good signal, and I'm not crazy about having to add a bunch of `NotImplemented` methods to interfaces when I even need a time aligning feature anyway.

**Comment by CodyCBakerPhD** [comment]
> Can be bypassed easily by defining an empty method.

Or by inheriting from a more aptly named base interface that does that for them


> Introduces more frictions in development. It increases development time by imposing additional requirements on developers. In other words, anyone creating a new interface now needs to implement these methods and consider their implications, which incurs an upfront cost.

The friction was always there. It was just hidden; we used to perform alignment in `__init__` or `run_conversion` methods in an ad-hoc way per interface. And for anyone unaware that this is a requirement of NWB, they would not be aware of any of this and ultimately create an NWB file that wasn't temporally aligned, which wouldn't be caught until the end of the process (uploaded to DANDI and contacted by us) which is arguably the worst time to encounter the issue.

The point of making these abstract is to bring this issue into the light from the very start.


> Breaks compatibility with previous code.

Previous code in past projects should have always used strict version pinning to freeze the environment to ensure stability. 

There are actually many reasons why a given project not using strict version pinning wouldn't work in a modern installation environment (usually changes to SI and other more subtle changes to NeuroConv that have broken compatibility, etc.); this particular change is just one of many potential things that could cause a break.

There's also a more central reason related to previous code; previous interfaces used ad-hoc methods of alignment in either `__init__` or `run_conversion` of an interface or, more commonly, in those methods of the `NWBConverter` itself. Those past approaches would not be in agreement with the new workflow, and might not even be guaranteed to work in conjunction with the new core interfaces to that extent either.

As such, previous pipelines would need some refactoring to use the latest NeuroConv features, which is expected with or without this change.



> For me, a NotImplementedError when a user tried to time-align is a pretty good signal,

The problem with this is at no point in any pipelines are we enforcing these methods to be used. If someone outside the team, unaware of this feature, put together their own pipelines for an NWB conversion, they might never know these methods exist, or how to use them, or why to use them.

They could simply make their custom interfaces, glue them together with a bunch of core interfaces for their data streams, and be happy creating an NWB file even though none of the data streams would be in sync with each other.

But if the methods are abstract, they get warned about it very early on in an attention-grabbing way to emphasize the importance of these methods, especially for custom behavior data.

**Comment by CodyCBakerPhD** [comment]
I would be open to a soft deprecation cycle however, where these do not start off as abstract but then become abstract at a specified point in time with plenty of warning leading up to it that people need to start specifying if data is pre-aligned, or if not, how to access the time information so that it can be

Would that satisfy the question for now @h-mayorquin?

**Comment by h-mayorquin** [comment]
> Or by inheriting from a more aptly named base interface that does that for them (the TemporallyAlignedDataInterface)

Is this a thing that we have already? I could not find it on the repo.

@CodyCBakerPhD 
OK. Would it be fair to describe your position as you want to put a bump in the road of users to educate them? I disagree with that philosophy of usability (as we have talked about this before) but let's argue from within in.  As a preliminar framework there are three things:

1) Where is the bump (e.g. `initialization` vs `run_time` or somewhere else)
2) How strong is the bump  (a warning, an assertion, forcing you to overide a parameter or implement a class).
3) How does the bump educates?

Let's say that we agree with the current framework: things should be done at `__init__` and that the bump is such that you can't use the interface unless you overide the methods. Even if I think we can do better, my biggest concern remains 3. Look at how my experience looked like:

I tried to implement a new BehaviorInterface without those methods and got the following assertion:

```python
TypeError: Can't instantiate abstract class HuzsarBehaviorInterface 
with abstract methods align_timestamps, get_original_timestamps, get_timestamps
```

I don't feel this is very education or useful at the moment. The only reference to alignment there is that there is a method called `align_timestamps` but .. what do you do with them? Do we have current examples that we can point users on how to implement and use them? When will they use those methods? I am an experience users of the library and it was not obvious to me at first glance. I had to go and look at how they are implemented for spikeinterface objects to get an idea (And I am not sure I understand them fully now). 

In summary, I think that we need a better way of educating users on how to use "the neuroconv way of alignment". If we are going to force them to use that framework in order to user our intefaces I think we should offer at least some documentation and examples. That is, I disagree with making an education bump that is this strong without good education materials and resources.

**Comment by bendichter** [comment]
Parent classes are usually more general, and subclassses are more specific. I think it's a bit of an anti-pattern to define `DataInterface` to require methods and then have those made unnecessary by a child class. How about the reverse?:

```python
class TemporallyAlignableDataInterface(BaseDataInterface):
    @abstractmethod
    def get_original_timestamps(self) -> np.ndarray:
        """
        Retrieve the original unaltered timestamps for the data in this interface.

        This function should retrieve the data on-demand by re-initializing the IO.

        Returns
        -------
        timestamps: numpy.ndarray
            The timestamps for the data stream.
        """
        raise NotImplementedError(
            "Unable to retrieve the original unaltered timestamps for this interface! "
            "Define the `get_original_timestamps` method for this interface."
        )

    @abstractmethod
    def get_timestamps(self) -> np.ndarray:
        """
        Retrieve the timestamps for the data in this interface.

        Returns
        -------
        timestamps: numpy.ndarray
            The timestamps for the data stream.
        """
        raise NotImplementedError(
            "Unable to retrieve timestamps for this interface! Define the `get_timestamps` method for this interface."
        )

    @abstractmethod
    def align_timestamps(self, aligned_timestamps: np.ndarray):
        """
        Replace all timestamps for this interface with those aligned to the common session start time.

        Must be in units seconds relative to the common 'session_start_time'.

        Parameters
        ----------
        aligned_timestamps : numpy.ndarray
            The synchronized timestamps for data in this interface.
        """
        raise NotImplementedError(
            "The protocol for synchronizing the timestamps of this interface has not been specified!"
        )
```

Then all align-able interfaces, e.g. `RecordingExtractorInterface` etc. will inherit from `TemporallyAlignableDataInterface`. Any custom interfaces would continue to inherit from `BaseDataInterface`. The fact that these interfaces do not inherit from `TemporallyAlignableDataInterface` would indicate explicitly that they are not temporally align-able. This (I think) satisfy @CodyCBakerPhD 's desire for explicitness of which interfaces do and don't support alignment while maintaining the custom classes defined previously. Thoughts on this?

**Comment by CodyCBakerPhD** [comment]
@h-mayorquin Thanks for all the feedback, this is helpful to get an outside perspective on

>OK. Would it be fair to describe your position as "you want to put a bump in the road of users to educate them about alignment"? I disagree with that philosophy of usability (as we have discussed before when discussing about the schema) but let's argue from within in. As a preliminary framework here are three points:

Yep. I know it's not great for usability of NeuroConv (the GUIDE is of course the ultimate answer to this however) but it is great for the usability of NWB itself; most users that create an NWB file (with or without NeuroConv) and upload it to DANDI have no idea that all time references have to be with respect to the session start time. It's a Best Practice, but we know that most people don't bother reading through all of those before going through the process.

As such, there are no bumps in any road ATM to draw users attention to such a critical detail. PyNWB cannot tell and won't complain; there are 1 or 2 indirect checks in the NWB Inspector, but it's a fairly difficult thing to determine programmatically after the file is created.

So this is the only way I can think of to draw users attention until we add a page on it to the GUIDE workflow (at least several months away from that feature if not more).

Do you have any better ideas? Without it, it basically means that even if you use NeuroConv, if you don't use it properly you still won't have a valid NWB file

> 1. Where is the bump (initialization vs run_time or somewhere else)
> 2. How strong is the bump (a warning, an assertion, forcing you to overide a parameter or implement a class, etc).
> 3. How does the bump educates?

1. `__init__`, as early as possible
2. Forcing to implement 3 methods
3. My hope is that they are at least curious enough to read the updated documentation on how to make a data interface (WIP) which will lead them to the documentation specifically about the variety of temporal alignment procedures (also WIP)


Your other questions would hopefully be satisfied by the documentation on the feature, which is still a work in progress and could use some help getting it finished

I'm OK with Ben's solution adding this as a pure feature for now, but I do think we eventually need to consider something stronger

**Comment by bendichter** [comment]
@CodyCBakerPhD how about also if `BaseDataInterface.__init__` includes something like:

```python
if not self.no_temporal_alignment and not isinstance(TemporallyAlignableDataInterface):
    warn.warning(f"{self.__name__} is not a temporally align-able interface, so it lacks the methods to synchronize this interface with other interfaces. If this is needed, please see the documentation for aligning interfaces here: [url].}")
```

**Comment by CodyCBakerPhD** [comment]
A soft warning like that could satisfy the concern; what do you think @h-mayorquin?

**Comment by h-mayorquin** [comment]
@CodyCBakerPhD 
> As such, there are no bumps in any road ATM to draw users attention to such a critical detail. PyNWB cannot tell and won't complain; there are 1 or 2 indirect checks in the NWB Inspector, but it's a fairly difficult thing to determine programmatically after the file is created.

Yeah, my initial instinct was that, this should be a an inspector issue but soon I realized it is not trivial. Maybe we should still brainstorm some heuristics in that front at some point.

I like @bendichter solution. It does not break backwards compatbility and also gives a warning to users about it that we can use to point them to the documentation on how to this.
